src/Symfony/Component/Security/Http/AccessToken/AccessTokenHandlerInterface.php

```diff
@@ -22,7 +23,9 @@
 interface AccessTokenHandlerInterface
 {
     /**
+     * @return UserBadge|string a UserBadge or a userIdentifier to retrieve a UserBadge from a {@see \Symfony\Component\Security\Core\User\UserProviderInterface}
+     *
      * @throws AuthenticationException
      */
-    public function getUserIdentifierFrom(#[\SensitiveParameter] string $accessToken): string;
+    public function getUserBadgeFrom(#[\SensitiveParameter] string $accessToken): UserBadge|string;
```

Comment:
I'm not a big fan of returning a string or an object. What about always returning an object and have a `NotLoadedUserBadge` that would just have the user identifier which would need then to be converted to a "fully loaded" user badge?

Reply:
This is great idea that I didn't even think about yet!

The `UserBadge` is actually meant for non-loaded users (this is what the user loader function is doing). E.g. `new UserBadge($userIdentifier)` is the main use-case (internally, [a listener](https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/Security/Http/EventListener/UserProviderListener.php) then configures the user provider as loader for this badge).

So we can make this method always return the `UserBadge` and leave the implementation 2 choices:

```
// default implementation, loads the user identifier from the database
public function getUserBadgeFrom(#[\SensitiveParameter] string $accessToken): UserBadge
{
    $userIdentifier = ...;

    return new UserBadge($userIdentifier);
}

// self-contained implementation, create a user on the fly
public function getUserBadgeFrom(#[\SensitiveParameter] string $accessToken): UserBadge
{
    $user = User::fromAccessToken($accessToken);

    return new UserBadge($user->getUserIdentifier(), fn () => $user);
}
```

Reply:
I like it too. Would the `AccessTokenAuthenticator` then do `if ($userBadge->getUserLoader() === null) { $userBadge->setUserLoader($this->userProvider->loadUserByIdentifier(...)) }` ?

Reply:
Ah, yes indeed. In this case, the authenticator should set the user provider if the loader is empty.